src/sys/petsclib.cxx

```diff
@@ -36,6 +52,22 @@ PetscLib::~PetscLib() {
   }
 }
 
+KSP PetscLib::createKSPWithOptions(MPI_Comm& comm) {
+  KSP ksp;
```

Comment:
I find this name slightly confusing to start with -- I'm expecting this to set the actual options used, whereas it only sets the prefix such that future calls to set from options will find the right values. 

Would it make sense to pass the relevant section in here and then register the options with petsc at that point. We could either pass through the prefix or allow the section to provide a special prefix value?

Could we also call set from options here as well -- this would presumably then set the defaults, the external code may then override this and then call set from options again as it currently does in order to allow command line arguments to override the internal setup?

Reply:
Good point, I didn't think that through. I'm not sure from the PETSc docs, but I think all the options-setting functions override anything set before, so you have to call `KSPSetFromOptions()` last to allow the 'options database' to set things at run-time. I don't think there's any benefit to calling it earlier.

I'd bundled the call to `KSPSetOptionsPrefix` in with `KSPCreate` without thinking too hard, just with the aim of not adding yet another function/method that each solver needs to call in the set-up. It might make more sense to instead put `KSPSetOptionsPrefix` and `KSPSetFromOptions` together in a method?

Reply:
Re: the method name - if we decide to keep the method as-is, would `createKSPWithPrefix` be better? (along with a much better docstring in the header to explain why it's called that)